lib/internal/modules/esm/formats.js

```diff
@@ -35,7 +38,32 @@ function mimeToFormat(mime) {
   return null;
 }
 
+/**
+ * For extensionless files in a `module` package scope, or a default `module` scope enabled by the `--experimental-type`
+ * flag, we check the file contents to disambiguate between ES module JavaScript and Wasm.
+ * We do this by taking advantage of the fact that all Wasm files start with the header `0x00 0x61 0x73 0x6d` (`_asm`).
+ * @param {URL} url
+ */
+function getFormatOfExtensionlessFile(url) {
+  if (!experimentalWasmModules) { return 'module'; }
+
+  const magic = new Uint8Array(4);
+  let fd;
+  try {
+    fd = openSync(url);
+    readSync(fd, magic, 0, 4); // Only read the first four bytes
```

Comment:
Can you add a todo to here? Optimization opportunity merge open and read C++ calls

Reply:
Or I can just make the change, if you can tell me what it should be? Or do we not have a `fs` method for “open and read” just yet?

Reply:
Aside, the comment has a bit of an "add 1 to i" vibe to it.

Reply:
I agree that this should likely be optimized somewhat, but there is no need to do it now.
I don't think there is a low level operation we can use for this.

Reply:
I'm thinking about having a single C++ function particularly for this operation that avoids having 3 C++ calls. I'm happy to follow up after this pull request.

Reply:
Sure, do you mind suggesting what comment should go here?

Reply:
`// TODO(@anonrig): Optimize the following my having a single C++ call`

Reply:
@anonrig if this will get moved to C++, it would be great to read the next dword and return its value.
I.e. in terms of js it's
```js
const u32 = readFirst8BytesAsUint32Array(file);
if (u32[0] !== 0x0061736d) {
  return -1; // special value if magic number doesn't match
} else {
  return u32[1]; // version number if it matches
}
```
This way, we'll also get wasm version which would be handy in future.

Reply:
This is definitely beyond me so I think it’s safe to say that this can be a follow-up 😄

Reply:
@LiviaMedeiros that's what I was aiming for. exactly!